### ISSUE TITLE
[Bash] Strict special parameter highlighting

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#184]
Current code treat all punctuation and white spaces as single character special parameter that will case bugs, like the one reported at https://github.com/zufuliu/notepad2/issues/684:
![image](https://user-images.githubusercontent.com/38994291/253325478-522c9abf-6432-4dae-bf03-8788fe23edb6.png)

`*`, `@`, `#`, `?`, `-`, `$`, `!` are supported in [bash](https://www.gnu.org/software/bash/manual/bash.html#Special-Parameters), [zsh](https://zsh.sourceforge.io/Doc/Release/Parameters.html#Parameters), [dash](https://man.archlinux.org/man/dash.1#Special_Parameters), and [ksh](https://man.archlinux.org/man/ksh.1#Parameter_Expansion.); [tcsh](https://man.archlinux.org/man/tcsh.1.en#Variable_substitution_without_modifiers) adds`%` and `<` (highlight `$<`  seems will cause other bug).

A simple fix is adding extra check after `scalar has no delimiter pair` comment: 
```cpp
// `setParam` will be replaced with `IsBashParameter()`.
if (!IsBashParameter(sc.chNext) && !IsSpecialParameter(sc.chNext)) {
	stylingInside = false;
}
```
https://github.com/ScintillaOrg/lexilla/blob/5f0c21517fa71a2f00c29fd5997538aed051c42c/lexers/LexBash.cxx#L362-L368

**Comment by zufuliu** [comment]
Another fix is adding 
```cpp
const CharacterSet setParamStart(CharacterSet::setAlphaNum, "\'\"`{[(_*@#?-$!");

if (sc.ch == '$' && setParamStart.Contains(sc.chNext)) {
	QuoteStack.Expand(sc, cmdState, stylingInside);
	continue;	
}
```

so special parameter list could be configuration.

**Comment by nyamatongwe** [comment]
Since the different shells have different choices, using a `CharacterSet` that could be modified by options appears sensible to me.

**Comment by zufuliu** [comment]
The patch, added `setParamStart` into `QuoteStackCls` to avoid code duplication.
[bash-param-0716.zip](https://github.com/ScintillaOrg/lexilla/files/12063061/bash-param-0716.zip)

**Comment by zufuliu** [comment]
Uploaded another fix at #186, which added property `lexer.bash.special.parameter` to set custom special parameters.
